### PR TITLE
[INTERNAL,FIX] Fixes initial draft of sequence_file_in.

### PR DESCRIPTION
**Pull request by smehringer**
this includes changes from #16 so it can be closed.

### CODE REVIEW

File: include/seqan3/io/sequence/sequence_file_format_fasta.hpp [hunk 1]
```diff
@@ -6,61 +43,103 @@
...
\ No newline at end of file
+template <typename seqs_type,
+          typename metas_type,
+          typename quals_type,
+          typename stream_type,
+          typename options_type>
+    requires sequence_of_sequence_concept<std::decay_t<seqs_type>> &&
+             sequence_of_sequence_concept<std::decay_t<metas_type>> &&
+             sequence_of_sequence_concept<std::decay_t<quals_type>>
+void sequence_file_format_fasta::read(seqs_type && seqs,
```
Comment (rrahn) [inline include/seqan3/io/sequence/sequence_file_format_fasta.hpp]: inline and put sequence name on separate line. Please do this for all interfaces

File: include/seqan3/io/sequence/sequence_file_in.hpp [hunk 2]
```diff
@@ -1,8 +1,52 @@
...
+// DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+// SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+// CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+// LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
+// OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH
+// DAMAGE.
+//
+// ==========================================================================
+// Authors: Svenja Mehringer, Temesgen H. Dadi, Jongkyu Kim
+//          <svenja.mehringer@fu-berlin.de> <temesgen.dadi@fu-berlin.de>
```
Comment (rrahn) [inline include/seqan3/io/sequence/sequence_file_in.hpp]: also put authors on different lines.

File: include/seqan3/io/sequence/sequence_file_in.hpp [hunk 3]
```diff
@@ -1,8 +1,52 @@
...
+
 #pragma once
 
 #include <string>
 #include <variant>
 #include <vector>
+#include <fstream>
+#include <functional>
+
+#include "../../container/concepts.hpp"
```
Comment (rrahn) [inline include/seqan3/io/sequence/sequence_file_in.hpp]: I suggest absolut paths.

File: include/seqan3/io/sequence/sequence_file_in.hpp [hunk 4]
```diff
@@ -14,155 +58,123 @@ namespace seqan3
 template <typename t>
 concept bool sequence_file_in_traits_concept = requires (t v)
 {
-    t::stream_type;
-    t::valid_formats;
-    requires detail::meets_concept_sequence_file_format<0, typename t::valid_formats>();
-
+    typename t::stream_type;
+    typename t::valid_format_types;
+    requires detail::meets_sequence_file_format_concept<typename t::valid_format_types>(
```
Comment (rrahn) [inline include/seqan3/io/sequence/sequence_file_in.hpp]: add empty line between type requirements and other components

File: include/seqan3/io/sequence/sequence_file_in.hpp [hunk 5]
```diff
@@ -14,155 +58,123 @@ namespace seqan3
...
-                                       sequence_file_format_genbank<stream_type>,
-                                       sequence_file_format_raw<stream_type>>;
-    static constexpr std::vector<std::pair<std::string, void>> valid_compression_formats{};
+    using stream_type = std::ifstream;
+    using valid_format_types = std::variant<sequence_file_format_fasta/*,
+                                       sequence_file_format_fastq,
+                                       sequence_file_format_embl,
+                                       sequence_file_format_genbank,
+                                       sequence_file_format_raw*/>;
+    using valid_compressions = std::variant<char/*specify compression formats*/>;
```
Comment (rrahn) [inline include/seqan3/io/sequence/sequence_file_in.hpp]: you can specify it as: 
```std::variant<decltype(std::ignore)>``` 😏

File: include/seqan3/io/sequence/sequence_file_in.hpp [hunk 6]
```diff
@@ -14,155 +58,123 @@ namespace seqan3
...
 
 // ==================================================================
 // sequence_file_in
 // ==================================================================
 
-template <typename sequence_file_in_traits = sequence_file_in_traits_default>
+template <typename sequence_file_in_traits = sequence_file_in_default_traits>
     requires sequence_file_in_traits_concept<sequence_file_in_traits>
-class sequence_file_in
+class sequence_file_in : public detail::file_base<sequence_file_in_traits>
```
Comment (rrahn) [inline include/seqan3/io/sequence/sequence_file_in.hpp]: protected inheritance?

File: include/seqan3/io/sequence/sequence_file_in.hpp [hunk 7]
```diff
@@ -14,155 +58,123 @@ namespace seqan3
...
-    using sequence_file_in_traits::valid_compression_formats; // = std::vector<std::pair<std::string, t>>
-    using sequence_file_in_traits::valid_formats;             // = std::variant<gzip_compressor, sequence_file_format_bam>
-
-    // constructor with arg
-    sequence_file_in(std::string const & _file_name);
-
-    // copy construction and assignment are deleted
-    // implicitly because we don't want multiple access to file
+    /* types */
+    using stream_type = typename detail::file_base<sequence_file_in_traits>::stream_type;
```
Comment (rrahn) [inline include/seqan3/io/sequence/sequence_file_in.hpp]: it suffices to say ```using detail::file_base<sequence_file_in_traits>::stream_type;```

File: include/seqan3/io/sequence/sequence_file_in.hpp [hunk 8]
```diff
@@ -14,155 +58,123 @@ namespace seqan3
...
+    using valid_format_types = typename detail::file_base<sequence_file_in_traits>::valid_format_types;
+
+    /* constructors */
+    sequence_file_in(std::experimental::filesystem::path _file_name) :
+        detail::file_base<sequence_file_in_traits>(_file_name) {};
```
Comment (rrahn) [inline include/seqan3/io/sequence/sequence_file_in.hpp]: move the _file_name in the constructor call of the base class.
declare the constructor ```explicit```.

File: include/seqan3/io/sequence/sequence_file_in.hpp [hunk 9]
```diff
@@ -14,155 +58,123 @@ namespace seqan3
...
-    std::string file_name;
-    stream_type stream;
-    valid_formats format;
+    /* member functions */
+    // TODO make the requirements stricter
+    template <typename sequence_type, typename meta_type, typename qual_type>
+        requires sequence_concept<std::decay_t<sequence_type>> &&
+                 sequence_concept<std::decay_t<meta_type>> &&
+                 sequence_concept<std::decay_t<qual_type>>
+    inline void read(sequence_type && seq,
```
Comment (rrahn) [inline include/seqan3/io/sequence/sequence_file_in.hpp]: you define but not declare the function ```inline```, according to https://isocpp.org/wiki/faq/inline-functions.

File: include/seqan3/io/sequence/sequence_file_in.hpp [hunk 10]
```diff
@@ -14,155 +58,123 @@ namespace seqan3
...
+             sequence_concept<std::decay_t<meta_type>> &&
+             sequence_concept<std::decay_t<qual_type>>
+inline void sequence_file_in<sequence_file_in_traits>::read(sequence_type && seq,
+                                                            meta_type && meta,
+                                                            qual_type && qual)
 {
-    // open stream
-    stream.open(_file_name, std::ios::binary);
+    if (format.valueless_by_exception())
+        throw "format not set!";
```
Comment (rrahn) [inline include/seqan3/io/sequence/sequence_file_in.hpp]: Can you add a todo, that we have to change the exception. 
actually we do not want to throw arbitrary types but real exception types.

File: include/seqan3/io/detail/file_base.hpp [hunk 11]
```diff
@@ -0,0 +1,126 @@
...
+public:
+    /* types */
+    using stream_type = typename file_base_traits::stream_type;
+    using valid_format_types = typename file_base_traits::valid_format_types;
+
+protected:
+
+    /* constructors */
+    // constructor with arg
+    file_base(std::experimental::filesystem::path _file_name)
```
Comment (rrahn) [inline include/seqan3/io/detail/file_base.hpp]: explicit 😏

File: include/seqan3/io/sequence/sequence_file_format.hpp [hunk 12]
```diff
@@ -36,49 +36,46 @@
 
 #pragma once
 
-#include <seqan3/alphabet/nucleotide/dna4_sequence.hpp>
-#include <seqan3/io/sequence/sequence_file_in.hpp>
 #include <string>
 #include <fstream>
+#include "../../alphabet/nucleotide/dna4_container.hpp"
```
Comment (rrahn) [inline include/seqan3/io/sequence/sequence_file_format.hpp]: please insert empty line between different library includes

File: include/seqan3/io/sequence/sequence_file_format_fasta.hpp [hunk 13]
```diff
@@ -6,61 +43,105 @@
...
-    sequence_file_format_fasta(sequence_file_in &&) = delete;
-    sequence_file_format_fasta & operator=(sequence_file_in &&) = delete;
-
-    static std::vector<std::string> file_extensions
-    {
-        { "fasta" },
-        { "fa" }
-        // ...
-    };
+    static inline std::vector<std::string> file_extensions {{".fasta"},{".fa"}};
```
Comment (rrahn) [inline include/seqan3/io/sequence/sequence_file_format_fasta.hpp]: space -> ```}, {".fa"}```

File: include/seqan3/io/sequence/sequence_file_in.hpp [hunk 14]
```diff
@@ -14,155 +58,127 @@ namespace seqan3
 template <typename t>
 concept bool sequence_file_in_traits_concept = requires (t v)
```
Comment (rrahn) [inline include/seqan3/io/sequence/sequence_file_in.hpp]: please put this concept in ```seqan3::detail ``` namespace.

File: include/seqan3/io/detail/file_base.hpp [hunk 15]
```diff
@@ -0,0 +1,126 @@
...
+// ==================================================================
+// file_base
+// ==================================================================
+
+template <typename file_base_traits>
+class file_base
+{
+public:
+    /* types */
+    using stream_type = typename file_base_traits::stream_type;
```
Comment (smehringer) [inline include/seqan3/io/detail/file_base.hpp]: This does not work.

File: include/seqan3/io/sequence/sequence_file_in.hpp [hunk 16]
```diff
@@ -14,155 +58,127 @@ namespace seqan3
...
-protected:
-    ~sequence_file_in() = default;
-
-private:
-    /* file format */
-    std::string file_name;
-    stream_type stream;
-    valid_formats format;
+    /* member functions */
+    // TODO make the requirements stricter
```
Comment (smehringer) [inline include/seqan3/io/sequence/sequence_file_in.hpp]: I don't think so since the requires might be replaced by range concepts or else.

**Pull request closed without merging**